### ISSUE TITLE
Sequence classification

### ISSUE THREAD

**Issue by RDaneelOlivav** [numenta/nupic#2742]
I launch this problem: How can we classify different sequences based on the SP1-->TP1-->SP2-->TP2?
For example, if we have two diferent sequences ( ABC s1 and AAB s2) how can we know that when we input an A, we are in s1 and not in s2, and therefore output from A-->SP1-->TP1-->s1, so that the level 2 in the hierarchy gets as inputs s1 or s2?

What I have now is to try and use Clustering to group up sequences of paterns based on their probability of transition to other patterns. I'm testing clustering libs in python such as sklearn.cluster and I'm also looking in the TP class if we can access those probabilities of transition between patterns, which I suppose we can.

Does someone know a fast way of getting those probability transitions from a NIPIC TP?

**Comment by rhyolight** [comment]
Any help @scottpurdy or @chetan51 ?

**Comment by chetan51** [comment]
I'm not sure I fully understand the entire post, but I will answer the question at the end of the post:

> Does someone know a fast way of getting those probability transitions from a NIPIC TP?

You can add a CLAClassifier to the network, feeding in the output of the TP region to the classifier. The classifier will give you the probabilities of the next transition in the sequence.

**Comment by RDaneelOlivav** [comment]
Update:
For the moment I've managed to acces the probability transitions of the TP and create conectivity matrix from that to classify the learned patterns into clusters which basicaly are classifying each patterns into a sequence. 
Lets say we have the following patterns: A,B,C,D,E,F
And we learn three sequences: 
s1) ABCD
s2) EF
s3) AADC
Well for now I've managed to create a system that "more or less" can tell you when you have an A, if the input is s1 or s3, by clustering those patterns, and outputting those sequences to the next level in the HTM. 
My question is if there was something implemented in NUPIC that did that, whats suposedly HTMs should finaly do, because it makes no sense that we can learn patterens and sequences but cant output those sequences when we are predicting.

Any new info?

**Comment by RDaneelOlivav** [comment]
Really? In the hotgym there is an example of sequence clasification??! Where? I've had a look at it and can't find where are you obtaining those sequences.

**Comment by chetan51** [comment]
I was just responding to your comment about being able to "output those sequences when we are predicting". I realize now you meant "output the labels of the sequences...", and for that there aren't any examples yet (but we are working on it in research). Perhaps @ywcui1990 or @marionleborgne will have more info on the state of sequence classification.

**Comment by marionleborgne** [comment]
Hi @RDaneelOlivav :) You might me interested to check out this project about sensor data sequence classification https://github.com/numenta/nupic.research/tree/master/projects/classification. As @chetan51 is saying, this is using the CLAClassifier on top of the TM to output class labels.

This lives in nupic.research so it is still fairly experimental. So let me know if you have comments of questions. Have a good weekend!

**Comment by RDaneelOlivav** [comment]
Thanks @chetan51 and @marionleborgne ;). I've managed alreadey to classify it more or less with Agglomerative clustering but I'll check out what you gave me @marionleborgne ;), because it sure is better than my system. It seems you arent useing the raw SP-TP structure but more the swarm-model structure right?

**Comment by ywcui1990** [comment]
@RDaneelOlivav We are using the raw SP->TM->TP network for the experiment. The HTM temporal memory learns sequences with the SP->TM structure, and then we take a union of the active cells in TM for each sequence and classify that. This is the basic idea, although there are many variants around it. As Marion said this is still fairly experimental.

**Comment by rhyolight** [comment]
@RDaneelOlivav Since this is more of a discussion than a feature request or bug report, do you mind if I close this issue? Or did I misunderstand?
